src/Collateral.sol

```diff
@@ -0,0 +1,202 @@
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {TAPVerifier} from "./TAPVerifier.sol";
+import {AllocationIDTracker} from "./AllocationIDTracker.sol";
+
+import "forge-std/console.sol";
+
+/**
+ * @title Collateral
+ * @dev This contract allows `senders` to deposit collateral for specific `receivers`,
+ *      which can later be redeemed using Receipt Aggregate Vouchers (`RAV`) signed
+ *      by an authorized `signer`. `Senders` can deposit collateral for `receivers`,
+ *      authorize `signers` to create signed `RAVs`, and withdraw collateral after a
+ *      set `thawingPeriod` number of blocks. `Receivers` can redeem signed `RAVs` to
+ *      claim collateral.
+ * @notice This contract uses the `TAPVerifier` contract for recovering signer addresses
+ *         from `RAVs`.
+ */
+contract Collateral {
+    struct CollateralAccount {
+        uint256 balance; // Total collateral balance for a sender-receiver pair
+        uint256 amountThawing; // Amount of collateral currently being thawed
+        uint256 thawEnd; // Block number at which thawing period ends
+    }
+
+    // Stores how much collateral each sender has deposited for each receiver, as well as thawing information
+    mapping(address => mapping(address => CollateralAccount)) private collateralAccounts;
+    // Map of authorized signers to which sender they are authorized to sign for
+    // key: signer address, value: sender address
+    mapping(address => address) private authorizedSigners;
+
+    // The ERC20 token used for collateral
+    IERC20 public immutable collateralToken;
+
+    // The contract used for verifying receipt aggregate vouchers
+    TAPVerifier public immutable tapVerifier;
+
+    // The contract used for tracking used allocation IDs
+    AllocationIDTracker public immutable allocationIDTracker;
+
+    // The duration in which collateral funds are thawing before they can be withdrawn
+    uint256 public immutable thawingPeriod;
+
+    /**
+     * @dev Emitted when collateral is deposited for a receiver.
+     */
+    event Deposit(address indexed sender, address indexed receiver, uint256 amount);
+
+    /**
+     * @dev Emitted when collateral is redeemed by a receiver.
+     */
+    event Redeem(address indexed receiver, uint256 amount);
+
+    /**
+     * @dev Emitted when a thaw request is made for collateral.
+     */
+    event ThawRequest(address indexed sender, address indexed receiver, uint256 amount, uint256 thawEnd);
+
+    /**
+     * @dev Emitted when thawed collateral is withdrawn by the sender.
+     */
+    event Withdraw(address indexed sender, address indexed receiver, uint256 amount);
+
+    /**
+     * @dev Emitted when a signer is authorized to sign RAVs for a sender.
+     */
+    event AuthorizeSigner(address indexed signer, address indexed sender);
+
+    constructor(address collateralToken_, address tapVerifier_, address allocationIDTracker_, uint256 thawingPeriod_) {
```

Comment:
Total nit but we usually use leading `_`s so `_collateralToken` instead of `collateralToken_`.

Reply:
Ah that's excellent. Looks like it's us who are out of spec then 😵‍💫